### DISCUSSION TITLE
Error: Unable to open database file: out of memory (14)

### DISCUSSION THREAD

**Discussion by ddbrn** [pocketbase/pocketbase#4624] [Q&A] [answered]
I am trying to install PocketBase on my M2 MacBook Air but after executing `./pocketbase serve` I get: `Error: Unable to open database file: out of memory (14)`

**Comment by ganigeorgiev** [comment]
I don't know how to reproduce it as I don't have a Mac.

Double check that you use the correct binary for your platform (M2 is ARM based so the [darwin_arm64](https://github.com/pocketbase/pocketbase/releases/download/v0.22.6/pocketbase_0.22.6_darwin_arm64.zip) binary should be used).

**Reply by ddbrn** [reply]
Thank you for your reply. 
I double-checked, and I am using the correct binary, but I am still getting the same error :(

**Reply by ganigeorgiev** [reply]
Is the `pb_data` directory and more specifically the `pb_data/data.db` file created?
If not then it is possible to be caused by a permission issue, I'm not sure.

**Reply by ddbrn** [reply]
Yeah, both have been created.

**Reply by ganigeorgiev** [reply, marked as answer]
Then I'm out of ideas.

It is possible to be a platform specific issue with the pure Go SQLite driver but as of now no one else has reported it and I have no easy way to verify it.

An alternative to the prebuilt executable is to try building PocketBase on your own:

0. [Download Go 1.21+](https://go.dev/doc/install) (_if you haven't already_)
1. Clone/download the repo:
   ```sh
   git clone https://github.com/pocketbase/pocketbase.git
   ```
2. Navigate to `example/base`
3. Run `CGO_ENABLED=0 go build`
4. Start the created executable: `./base serve`

If it still doesn't work, try to replace step3 with `CGO_ENABLED=1 go build`.

_Edit: With v0.23+ for the CGO version you have to enable the driver explicitly as in https://pocketbase.io/docs/go-overview/#github-commattngo-sqlite3._

**Reply by ganigeorgiev** [reply]
Just to rule it out - the location of the executable and `pb_data` is a regular and writable directory, aka. it is NOT a network drive or some other filesystem virtualization, right?

**Reply by ddbrn** [reply]
Yes, it is just a regular and writable directory. However, building PocketBase on my own seems to be working for me. 
Thank you!

**Reply by ganigeorgiev** [reply]
If you used `CGO_ENABLED=0` then the prebuilt executable should be identical to your manually generated one, aka. this means that something else is most likely going on in your environment but I have no idea what that could be.

If it works only with `CGO_ENABLED=1` then most likely it is a platform specific bug/incompatibility with the pure Go driver but again no else so far has reported it neither in the PocketBase repo, nor in https://gitlab.com/cznic/sqlite. If someone else is having the same issue, please consider reporting it with more details about your environment to attempt to reproduce it.

**Reply by ddbrn** [reply]
It worked with `CGO_ENABLED=0` for me.

**Comment by devjume** [comment]
I had same error when I tried to run customized/extended version of Pocketbase in Docker Compose with mounted volume.
Mounted volume triggered the error for me.  Here are my notes about the issue for anyone who might face same error.

**Solution when using Pocketbase with Docker:**
Create new user in your Dockerfile and give it permission to folder where Pocketbase executable is.
This user runs the Pocketbase executable and will have permissions that will work with mounted volume.

Here are relevant commands from my Dockerfile:
```
# Base image golang:1.22.4-alpine3.20

# Build the application
RUN CGO_ENABLED=0 GOOS=linux go build -a -o /pocketbase

# Create the home directory for the new nonroot user.
RUN mkdir -p /home/nonroot

# Create non-root user and group so our program doesn't run as root.
# -S = Create a system user or group
# -D = Don't assign a password, so cannot login
# -h = Home directory
# -G = Add user to group
RUN addgroup -S nonroot && adduser -S -D -h /home/nonroot -G nonroot nonroot

# Set the home directory to our nonroot user's home.
ENV HOME=/home/nonroot
ENV APP_HOME=/home/nonroot/app

RUN mkdir $APP_HOME
WORKDIR $APP_HOME

# Copy the built application from the build stage
COPY --from=build-stage /pocketbase $APP_HOME/pocketbase

# Chown all the files to the app user.
RUN chown -R nonroot:nonroot $APP_HOME
```

Second options is to use `root` user to run the executable but that's not recommend.

Refences where I came up with the solution and that mentioned the root user option.
https://gist.github.com/developer-guy/52965048c33be0fb2831dc284c50aebc
https://github.com/tailscale/golink/issues/6

**Comment by jhass** [comment]
Adding my experience since googling the error message can get people land here.

I'm deploying Pocketbase inside a Kubernetes cluster using Longhorn for storage, which internally means the volume might be mounted via NFS into the container. I saw this error while for some reason the NFS server crashed and the directory listing for the data directory came back empty.

**Reply by plaes** [reply]
The issue seem to be that errno from SQLite is interpreted wrong. Error 14 means that file cannot be opened: https://sqlite.org/rescode.html#cantopen

As `CGO_ENABLED=1` is using C -> Go translation directly, it's getting the error code right. Therefore pure Go-based somehow maps that error code (14) to wrong code:
https://sqlite.org/c3ref/c_abort.html

**Reply by ganigeorgiev** [reply]
@plaes You are commenting to an old discussion and I'm not sure that what you are saying is true. In normal operations it is not expected the SQLite error to surface and in addition to the SQLite busy_timeout we have auto-fail retry on application level (we don't rely on the error code).

In any case, I'll lock this discussion as it starting to become too off-topic and spammy.

If anyone is experiencing "Unable to open database file: out of memory", please open a new discussion but try to provide clear reproducible steps so that me and other users can assist you.
